<a href="https://colab.research.google.com/github/avi-dot-ai/w1-1/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. My lane (or freestyle) and why**

**Lane: Ranking Signal Analysis.**
I chose the Ranking Signal Analysis lane because During the initial exploration, I found that search volume has almost no linear relationship with impressions (correlation ≈ 0.001), and average position also showed only a weak linear relationship with CTR (correlation ≈ -0.073). These observations suggest that ranking performance is influenced by multiple interacting factors rather than a single metric. Therefore, I chose the Ranking Signal Analysis lane to investigate which combinations of page-level signals are most useful for understanding and improving search performance. And understanding the factors that influence search rankings has broad applications in search engines, recommendation systems, and AI-driven products. My goal is to identify which measurable signals are most strongly associated with higher-ranking pages and translate those findings into evidence-backed recommendations. This project combines exploratory data analysis, statistical reasoning, feature engineering, and machine learning, making it a strong opportunity to develop practical ML skills while solving a real business problem.

# **2. The question: decision, action, cost of a wrong call**

**Research Question:**

Which page-level signals are most strongly associated with higher search rankings, and which of these should be prioritized to improve organic search performance?

**Decision:**

Determine which ranking signals deserve the highest priority for optimization.

**Action:**

SEO specialists or content teams can optimize pages by improving the most influential signals identified through the analysis.

**Unit of Analysis:**

One webpage (one URL).

**Output:**

A ranked report of important ranking signals with supporting evidence, feature importance analysis, and actionable recommendations.

**Cost of a Wrong Recommendation:**

If unimportant signals are prioritized, teams may spend significant engineering and content resources on changes that do not improve rankings, resulting in wasted time, higher costs, and missed opportunities to improve search visibility.

**Why ML Helps:**

Ranking performance depends on many interacting variables that are difficult to evaluate manually. Data analysis and machine learning can identify meaningful patterns, quantify relationships between features and rankings, and prioritize the factors most likely to influence performance. The objective is to support better decisions rather than replace human judgment.

# **3. Quick look at the data (2-3 real numbers)**

In [2]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
elif os.path.basename(os.getcwd()) == "notebooks":
    os.chdir("..")  # move from notebooks/ to the repo root

print("Working dir:", os.getcwd())
assert os.path.exists("data/raw/content_refresh_anonymized.csv"), "starter CSV not found — are you at the repo root?"
print("Starter data found. You're ready.")

Working dir: /content/flyrank-ml-internship-starter
Starter data found. You're ready.


In [6]:
import pandas as pd, numpy as np
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")
print(f"Total pages in dataset: {len(df):,}")

print(f"Average CTR: {df['ctr'].mean():.2f}%")

print(f"Average Position: {df['avg_position'].mean():.2f}")

print(f"Pages with impressions > 0: {(df['impressions_90d'] > 0).sum():,} "
      f"({100*(df['impressions_90d'] > 0).mean():.1f}%)")

print(f"Median Impressions (90d): {df['impressions_90d'].median():,.0f}")

Total pages in dataset: 30,000
Average CTR: 0.51%
Average Position: 16.34
Pages with impressions > 0: 30,000 (100.0%)
Median Impressions (90d): 731


* **The starter dataset contains 30,000 webpages, providing a sufficiently large sample for ranking analysis. The average page position is 16.34, while the average CTR is 0.51%. Additionally, 731 pages received median impressions during the last 90 days, indicating meaningful variation in search performance. These statistics suggest that the dataset contains enough diversity to investigate which page-level signals are associated with better rankings.**

# **4. Careful words: what I can and can't claim**

* **What I can claim:**

**The analysis identifies statistical relationships between ranking signals and search performance.
The results provide evidence-based recommendations for prioritizing optimization efforts.
Machine learning models can estimate feature importance and help rank influential signals.**

* **What I cannot claim:**

**This analysis cannot prove that any signal directly causes higher rankings.
It cannot reveal Google's ranking algorithm.
The findings are based only on the available dataset and may not generalize to every website or industry.
The recommendations support decision-making but should be validated through controlled experiments such as A/B testing.**

#5. Self Check

* **Selected one predefined lane**
* **Defined a clear research question**
* **Identified the decision and action**
* **Specified the unit of analysis**
* **Explained the output**
* **Described the cost of incorrect recommendations**
* **Explained why ML is useful**
* **Included 2–3 real numbers from the dataset**
* **Used careful, evidence-based language**